File: packages/icue_face_sdk/android-core/src/main/kotlin/school/icue/face/core/quality/IcueFaceQualityAssessor.kt
package school.icue.face.core.quality

import android.graphics.Bitmap
import android.graphics.Color
import com.google.mlkit.vision.face.Face

enum class FaceQualityStatus {
    OK,
    TOO_SMALL,
    BLURRY,
    EXTREME_POSE
}

data class FaceQualityResult(
    val status: FaceQualityStatus,
    val blurScore: Float,
    val isValid: Boolean,
)

object IcueFaceQualityAssessor {
    const val DEFAULT_MIN_FACE_SIZE = 24
    const val DEFAULT_BLUR_THRESHOLD = 80.0f
    const val MAX_YAW_ANGLE = 35.0f
    const val MAX_PITCH_ANGLE = 25.0f

    /**
     * Assesses the quality of a face crop using bounding box size, head pose angles,
     * and grayscale Laplacian variance blur estimation.
     */
    fun assessQuality(
        bitmap: Bitmap,
        face: Face,
        minFaceSize: Int = DEFAULT_MIN_FACE_SIZE,
        minBlurThreshold: Float = DEFAULT_BLUR_THRESHOLD,
    ): FaceQualityResult {
        val bbox = face.boundingBox
        if (bbox.width() < minFaceSize || bbox.height() < minFaceSize) {
            return FaceQualityResult(FaceQualityStatus.TOO_SMALL, 0f, false)
        }

        val yaw = face.headEulerAngleY
        val pitch = face.headEulerAngleX
        if (Math.abs(yaw) > MAX_YAW_ANGLE || Math.abs(pitch) > MAX_PITCH_ANGLE) {
            return FaceQualityResult(FaceQualityStatus.EXTREME_POSE, 0f, false)
        }

        val blurScore = calculateLaplacianVariance(bitmap)
        val faceWidth = bbox.width()
        val adaptiveScale = (faceWidth.toFloat() / 112f).coerceIn(0.25f, 1.0f)
        val effectiveBlurThreshold = minBlurThreshold * adaptiveScale

        if (blurScore < effectiveBlurThreshold) {
            return FaceQualityResult(FaceQualityStatus.BLURRY, blurScore, false)
        }

        return FaceQualityResult(FaceQualityStatus.OK, blurScore, true)
    }

    /**
     * Calculates the variance of Laplacian on the bitmap to estimate image sharpness.
     * Higher variance indicates sharper edges; low variance indicates blur.
     */
    fun calculateLaplacianVariance(bitmap: Bitmap): Float {
        val width = bitmap.width
        val height = bitmap.height
        if (width < 3 || height < 3) return 0f

        val pixels = IntArray(width * height)
        bitmap.getPixels(pixels, 0, width, 0, 0, width, height)

        // Convert to grayscale
        val gray = FloatArray(width * height)
        for (i in pixels.indices) {
            val p = pixels[i]
            val r = (p shr 16) and 0xff
            val g = (p shr 8) and 0xff
            val b = p and 0xff
            gray[i] = 0.299f * r + 0.587f * g + 0.114f * b
        }

        // Apply 3x3 Laplacian kernel: [[0, 1, 0], [1, -4, 1], [0, 1, 0]]
        var sum = 0.0
        var sqSum = 0.0
        var count = 0

        for (y in 1 until height - 1) {
            val rowOffset = y * width
            val prevRowOffset = (y - 1) * width
            val nextRowOffset = (y + 1) * width

            for (x in 1 until width - 1) {
                val center = gray[rowOffset + x]
                val up = gray[prevRowOffset + x]
                val down = gray[nextRowOffset + x]
                val left = gray[rowOffset + x - 1]
                val right = gray[rowOffset + x + 1]

                val laplacian = up + down + left + right - 4f * center
                sum += laplacian
                sqSum += laplacian * laplacian
                count++
            }
        }

        if (count == 0) return 0f

        val mean = sum / count
        val variance = (sqSum / count) - (mean * mean)
        return variance.toFloat().coerceAtLeast(0f)
    }
}

File: packages/icue_face_sdk/ios/Classes/IcueFaceQualityAssessor.swift
import Foundation
import UIKit
import MLKitFaceDetection

public enum IcueFaceQualityStatus: String {
    case ok = "OK"
    case tooSmall = "TOO_SMALL"
    case blurry = "BLURRY"
    case extremePose = "EXTREME_POSE"
}

public struct IcueFaceQualityResult {
    public let status: IcueFaceQualityStatus
    public let blurScore: Float
    public let isValid: Bool
}

public class IcueFaceQualityAssessor {
    public static let DEFAULT_MIN_FACE_SIZE: CGFloat = 24.0
    public static let DEFAULT_BLUR_THRESHOLD: Float = 80.0
    public static let MAX_YAW_ANGLE: Float = 35.0
    public static let MAX_PITCH_ANGLE: Float = 25.0

    /**
     * Assesses the quality of a face crop using bounding box size, head pose angles,
     * and grayscale Laplacian variance blur estimation with resolution-adaptive thresholding.
     */
    public static func assessQuality(
        image: UIImage,
        face: Face,
        minFaceSize: CGFloat = DEFAULT_MIN_FACE_SIZE,
        minBlurThreshold: Float = DEFAULT_BLUR_THRESHOLD
    ) -> IcueFaceQualityResult {
        let frame = face.frame
        if frame.width < minFaceSize || frame.height < minFaceSize {
            return IcueFaceQualityResult(status: .tooSmall, blurScore: 0.0, isValid: false)
        }

        if face.hasHeadEulerAngleY && abs(face.headEulerAngleY) > MAX_YAW_ANGLE {
            return IcueFaceQualityResult(status: .extremePose, blurScore: 0.0, isValid: false)
        }

        if face.hasHeadEulerAngleX && abs(face.headEulerAngleX) > MAX_PITCH_ANGLE {
            return IcueFaceQualityResult(status: .extremePose, blurScore: 0.0, isValid: false)
        }

        let blurScore = calculateLaplacianVariance(image: image)
        let faceWidth = Float(frame.width)
        let adaptiveScale = max(0.25, min(1.0, faceWidth / 112.0))
        let effectiveBlurThreshold = minBlurThreshold * adaptiveScale

        if blurScore < effectiveBlurThreshold {
            return IcueFaceQualityResult(status: .blurry, blurScore: blurScore, isValid: false)
        }

        return IcueFaceQualityResult(status: .ok, blurScore: blurScore, isValid: true)
    }

    /**
     * Calculates variance of 3x3 Laplacian operator on grayscale pixel buffer.
     */
    public static func calculateLaplacianVariance(image: UIImage) -> Float {
        guard let cgImage = image.cgImage else { return 0.0 }
        let width = cgImage.width
        let height = cgImage.height
        if width < 3 || height < 3 { return 0.0 }

        var rawBytes = [UInt8](repeating: 0, count: width * height * 4)
        guard let context = CGContext(
            data: &rawBytes,
            width: width,
            height: height,
            bitsPerComponent: 8,
            bytesPerRow: width * 4,
            space: CGColorSpaceCreateDeviceRGB(),
            bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
        ) else { return 0.0 }

        context.draw(cgImage, in: CGRect(x: 0, y: 0, width: width, height: height))

        // Convert to grayscale
        var gray = [Float](repeating: 0.0, count: width * height)
        for i in 0..<(width * height) {
            let r = Float(rawBytes[i * 4])
            let g = Float(rawBytes[i * 4 + 1])
            let b = Float(rawBytes[i * 4 + 2])
            gray[i] = 0.299 * r + 0.587 * g + 0.114 * b
        }

        var sum: Double = 0.0
        var sqSum: Double = 0.0
        var count: Int = 0

        for y in 1..<(height - 1) {
            let rowOffset = y * width
            let prevRowOffset = (y - 1) * width
            let nextRowOffset = (y + 1) * width

            for x in 1..<(width - 1) {
                let center = gray[rowOffset + x]
                let up = gray[prevRowOffset + x]
                let down = gray[nextRowOffset + x]
                let left = gray[rowOffset + x - 1]
                let right = gray[rowOffset + x + 1]

                let laplacian = Double(up + down + left + right - 4.0 * center)
                sum += laplacian
                sqSum += laplacian * laplacian
                count += 1
            }
        }

        if count == 0 { return 0.0 }

        let mean = sum / Double(count)
        let variance = (sqSum / Double(count)) - (mean * mean)
        return Float(max(0.0, variance))
    }
}
